include/task.h

```diff
@@ -15,51 +15,72 @@
 #include "reduction.h"
 #include "types.h"
 
-
 namespace celerity {
 
 class handler;
 
 namespace detail {
 
+	struct task_geometry {
+		int dimensions = 0;
+		range<3> global_size{1, 1, 1};
+		id<3> global_offset;
+		range<3> granularity{1, 1, 1};
+	};
+
 	class buffer_access_map {
 	  public:
-		void add_access(buffer_id bid, std::unique_ptr<range_mapper_base>&& rm) { m_accesses.emplace_back(bid, std::move(rm)); }
+		struct buffer_access {
+			buffer_id bid;
+			access_mode mode;
+			std::unique_ptr<range_mapper_base> range_mapper;
+		};
```

Comment:
`buffer_access` is often used outside the BAM, so it might make code simpler to move this struct to the enclosing scope.

Also consider initializing all members.

Reply:
I wouldn't have to spell it out everywhere if `emplace_back` would work... 😑 Done!